obci/devices/haptics/HapticsControl.py

```diff
@@ -0,0 +1,140 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Author:
+#      Marian Dovgialo <marian.dowgialo@gmail.com>
+#
+'''
+Contains HapticStimulator class to control FTDI FT2232HL in bitbang 
+mode as haptic stimulator.
+Tested on pyftdi 0.11.3 and pyusb 1.0.0.b2
+To work requires user to have access to usb device.
+On systems with udev:
+add file/etc/udev/rules.d/99-FTDI.rules
+with line (idVendor and idProduct might be differentfor your
+stimulator):
+SUBSYSTEM=="usb", ATTR{idVendor}=="0403", ATTR{idProduct}=="6010", MODE="666"
+'''
+
+from pyftdi.ftdi import Ftdi
+import threading
+import time
+
+VENDORID = 0x0403
+PRODUCTID = 0x6010
+DEFAULTINTERFACE = 2 # Interface to controll
+DEFAULTDIRECTION = 0xFF # All pins - output
+
+
+class HapticStimulator(object):
+    '''
+    Class to initialise and send commands to sensory stimulator built 
+    on FTDI FT2232HL with power trasistors board
+    ("fMRI Pneumatic 5V version") from Politechnika Warszawska
+    '''
+    def __init__(self, vendorid=VENDORID, productid=PRODUCTID,
+                        interface=DEFAULTINTERFACE):
+        '''
+        Initialises FTDI device as Haptic Stimulator
+        
+        :param vendorid: int or hex NOT string
+        gotten from lsusb for FTDI device
+        :param productid: int or hex NOT string
+        gotten from lsusb for FTDI device
+        :param interface: int - for tested board is always 2'''
+        self.ftdi = Ftdi()
+        self.ftdi.open_bitbang(vendorid, productid, interface, 
+                                direction = 0xFF)
+                                # direction FF - all ports output
+        self.lock = threading.Lock()
+        self.ftdi.write_data([0]) # turn off all channels
+        
+    def __del__(self):
+        self.close()
+    
+    def _turn_off(self, chnl):
+        '''
+        Will turn off selected channel, used by stimulate method on
+        on timer
+        
+        :param chnl: int - number of channel to turn off
+        '''
+        with self.lock:
+            apins = self.ftdi.read_pins()
+            activation_mask = ~(1 << (chnl-1)) & 0xFF #select only 
+                                                      #needed channel
+            self.ftdi.write_data([apins & activation_mask])
+    
+    def stimulate(self, chnl, time):
+        '''
+        Turn on stimulation on channel number chnl for time seconds
+        
+        :param chnl: integer - channel number starting from 1
+        :param time: float - stimulation length in seconds
+        '''
+        t = threading.Timer(time, self._turn_off, [chnl]) # we expect 
+                                                #short times ~ seconds
+                                                #handling timers 
+                                                #should not matter
+        t.daemon = True # timer thread will shutdown at main program
```

Comment:
This comment is not true - daemon threads are **abruply** stopped at shutdown. No destructor code is run.